### PR TITLE
Some fixes for custom allreduce kernels

### LINKED ISSUE #2728: Mixtral GPTQ with TP=2 not generating output
In the new vllm 0.3 release mixtral with gptq does not generate any output anymore. Loading the model works fine, when calling the `llm.generate` it gets stuck. 

```
from vllm import SamplingParams
import torch 

model_name = "TheBloke/Mixtral-8x7B-Instruct-v0.1-GPTQ"
llm = vllm.LLM(model=model_name, quantization="gptq", dtype=torch.float16, tensor_parallel_size=2, max_model_len=16000, revision="gptq-4bit-32g-actorder_True", gpu_memory_utilization=0.75, enforce_eager=False)
llm.generate(formatted_prompt[0], sampling_params=SamplingParams(temperature=0.1, max_tokens=100))
```

Currently using: 
- python 3.11.* 
- vllm-0.3.0+cu123 -> tried also the current git repo
- cuda 12.3

The worker seems to be stuck in the `llm_engine.step()` function / `_run_workers` call.

### LINKED ISSUE #2795: Distributed inference on multi machine (error Invalid peer device id)
I'm a newbie, and I'm running an example at https://docs.vllm.ai/en/latest/serving/distributed_serving.html locally with 2 machines, each with an RTX 3090 GPU. I changed tensor_parallel_size to 2 and model to "vinai/PhoGPT-4B".
On the head node, I run:
 NCCL_SOCKET_IFNAME=eth0 NCCL_DEBUG=INFO CUDA_VISIBLE_DEVICES=0 ray start --head. 
On the other nodes, I run:
 NCCL_SOCKET_IFNAME=eth0 NCCL_DEBUG=INFO CUDA_VISIBLE_DEVICES=0 ray start --address='10.0.0.1'.
Then, on the head node, when I run the example code: python main.py, I get the following error:
```
Traceback (most recent call last):
  File "/data2/bientd/vllm/test.py", line 25, in <module>
    llm = LLM(model="facebook/opt-13b", tensor_parallel_size=2,download_dir='/data2/bientd/')#,pipeline_parallel_size=3 don't support
  File "/data2/bientd/anaconda3/envs/vllm/lib/python3.9/site-packages/vllm/entrypoints/llm.py", line 109, in __init__
    self.llm_engine = LLMEngine.from_engine_args(engine_args)
  File "/data2/bientd/anaconda3/envs/vllm/lib/python3.9/site-packages/vllm/engine/llm_engine.py", line 356, in from_engine_args
    engine = cls(*engine_configs,
  File "/data2/bientd/anaconda3/envs/vllm/lib/python3.9/site-packages/vllm/engine/llm_engine.py", line 109, in __init__
    self._init_workers_ray(placement_group)
  File "/data2/bientd/anaconda3/envs/vllm/lib/python3.9/site-packages/vllm/engine/llm_engine.py", line 271, in _init_workers_ray
    self._run_workers("init_model")
  File "/data2/bientd/anaconda3/envs/vllm/lib/python3.9/site-packages/vllm/engine/llm_engine.py", line 983, in _run_workers
    driver_worker_output = getattr(self.driver_worker,
  File "/data2/bientd/anaconda3/envs/vllm/lib/python3.9/site-packages/vllm/worker/worker.py", line 87, in init_model
    init_custom_ar()
  File "/data2/bientd/anaconda3/envs/vllm/lib/python3.9/site-packages/vllm/model_executor/parallel_utils/custom_all_reduce.py", line 44, in init_custom_ar
    if not _can_p2p(rank, world_size):
  File "/data2/bientd/anaconda3/envs/vllm/lib/python3.9/site-packages/vllm/model_executor/parallel_utils/custom_all_reduce.py", line 137, in _can_p2p
    if not torch.cuda.can_device_access_peer(rank, i):
  File "/data2/bientd/anaconda3/envs/vllm/lib/python3.9/site-packages/torch/cuda/__init__.py", line 464, in can_device_access_peer
    raise AssertionError("Invalid peer device id")
AssertionError: Invalid peer device id
```

### PR DESCRIPTION
**Pull request by hanzhi713**
Recently, there are some reports of stuck generation or garbage text when custom all reduce is enabled. While I didn't manage to reproduce any issues on A30 and A100, I did find some potentially unsafe synchronizations and I attempt to fix them here. 

1. When using the signal flag, GPUs are writing to different bytes of the same 8-byte signal. Although the writes are __strong__ writes, they are not considered __morally strong__ according to the CUDA memory model because they didn't overlap completely. Hence, they are still considered as data races.
2. When using 2-stage allreduce or half butterfly allreduce, a `__threadfence_system` or a release-acquire pattern is needed to absolutely guarantee the visibility of other devices' write on the current device, which is missing from the current implementation. 

Related issues:
#2788 (garbage output when upgrading vllm from 0.2.7 -> 0.3.0)
#2742 (garbage output. Solved when disable_custom_all_reduce=True)
#2731 (one person reported that disable_custom_all_reduce=True solve generation hanging)

In this PR, I made the following changes
1. Use 8 uint32 instead of 8 bytes per signal per device.
2. simplify synchronization by only syncing blocks of the same index across GPUs. Some changes are made to ensure that each thread's read only depends on the writes from the thread of the same id in other devices.
3. add a `__threadfence_system` guarantee visibility of other device's writes when using 2-stage or half butterfly allreduce. Note that this adds a few microseconds of overhead. 
4. removed support for more than two PCIe-only GPUs because performance improvement is small.
5. add additional p2p checks to avoid buggy driver/hardware P2P support. Might fix #2728
6. add check for device count when running P2P test. Should fix #2795
7. <s>disable custom allreduce by default by setting the argument to True. User can explicitly opt-in by setting it to False.</s>

### CODE REVIEW

**Comment by hanzhi713** [comment]
@WoosukKwon This PR is ready for review. By the way, did any of you manage to produce custom all reduce stuck of generate garbage output?

**Comment by NikolaBorisov** [comment]
To reproduce I run vllm on 4x A100 80G SXM with CodeLlama 70b. I sent some requests like this:

```
for i in {0..100}; do curl "http://localhost:8000/v1/chat/completions" -H "Content-Type: application/json" -d '{
  "model": "codellama/CodeLlama-70b-Instruct-hf",
  "messages": [
    {"role": "user", "content": "Hello!"}
  ],
  "max_tokens":100
  }' & sleep 2; done
```

It usually gets stuck before request 50 for me.

**Comment by hanzhi713** [comment]
> To reproduce I run vllm on 4x A100 80G SXM with CodeLlama 70b. I sent some requests like this:
> 
> ```
> for i in {0..100}; do curl "http://localhost:8000/v1/chat/completions" -H "Content-Type: application/json" -d '{
>   "model": "codellama/CodeLlama-70b-Instruct-hf",
>   "messages": [
>     {"role": "user", "content": "Hello!"}
>   ],
>   "max_tokens":100
>   }' & sleep 2; done
> ```
> 
> It usually gets stuck before request 50 for me.

Did not manage to reproduce when `enforce_eager=True`. Custom allreduce is running fine I think

**Comment by hanzhi713** [comment]
@WoosukKwon This PR is ready for review.

**Comment by tdene** [comment]
Hi @hanzhi713 I'm about to start merging this into my build, tonight.    
Before I do though, in your comment from yesterday you mentioned    
> Did not manage to reproduce when `enforce_eager=True`.

Forgive my ignorance, but is the custom all reduce kernel not meant to work with CUDA graphs? If so, do you know which of the two features would make a bigger impact on latency?

**Comment by hanzhi713** [comment]
> Hi @hanzhi713 I'm about to start merging this into my build, tonight. Before I do though, in your comment from yesterday you mentioned
> 
> > Did not manage to reproduce when `enforce_eager=True`.
> 
> Forgive my ignorance, but is the custom all reduce kernel not meant to work with CUDA graphs? If so, do you know which of the two features would make a bigger impact on latency?

Custom allreduce does work with CUDA Graph. It's just that vLLM CUDA Graph previously had hanging issues so I didn't run the test with `enforce_eager=False`. Now since #2811 is merged, we would expect custom allreduce to run properly with CUDA graph enabled.

**Comment by hanzhi713** [comment]
Regarding the improvement in latency, this should vary depending on the model size and number of GPUs used. On average, CUDA graph can improve latency by about 10%, and custom allreduce can add another 10% on NVSwitch-enabled systems.

**Comment by hanzhi713** [comment]
@WoosukKwon can you take a look?

**Comment by tdene** [comment]
Just for reference, I have been using this PR merged on top of 37ca7f60 for about 2 weeks now, with no noticed issues.

**Comment by tdene** [comment]
Actually.

@hanzhi713 have you tested this PR with a MoE model like mixtral?    
I'm seeing    
```
custom_all_reduce.py:239] Registering 2 cuda graph addresses
[...]
File "/workspace/vllm/vllm/worker/worker.py", line 160, in warm_up_model 
  self.model_runner.capture_model(self.gpu_cache)
File "/usr/local/lib/python3.10/dist-packages/torch/utils/_contextlib.py", line 115, in decorate_context
   return func(*args, **kwargs)
File "/workspace/vllm/vllm/worker/model_runner.py", line 725, in capture_model
  graph_runner.capture(
[...]
File "/workspace/vllm/vllm/model_executor/models/mixtral.py", line 130, in forward
  final_hidden_states = fused_moe(hidden_states,
File "/workspace/vllm/vllm/model_executor/layers/fused_moe.py", line 276, in fused_moe
   gating_output.float(),  # TODO(woosuk): Optimize this.
RuntimeError: CUDA error: invalid device function
CUDA kernel errors might be asynchronously reported at some other API call, so the stacktrace below might be incorrect.
For debugging consider passing CUDA_LAUNCH_BLOCKING=1.
Compile with `TORCH_USE_CUDA_DSA` to enable device-side assertions.
```

**Comment by hanzhi713** [comment]
@tdene Hmm may the buffer isn't placed on the correct cuda device. Let me check tomorrow.

**Comment by hanzhi713** [comment]
@tdene I didn't manage to reproduce this error

**Comment by hanzhi713** [comment]
@tdene I merged this branch with the latest main. Can you rerun your test? Also, I find it strange to see `Registering 2 cuda graph addresses` because normally with cuda graph enabled, we should see at least a few hundred addresses, if not thousands.

File: vllm/config.py [hunk 1]
```diff
@@ -685,4 +676,4 @@ def _get_and_verify_max_len(
             " in model's config.json). This may lead to incorrect model "
             "outputs or CUDA errors. Make sure the value is correct and "
             "within the model context size.")
-    return int(max_model_len)
+    return int(max_model_len)
```
Comment (WoosukKwon) [inline vllm/config.py:679]:
```suggestion
    return int(max_model_len)

```

File: vllm/engine/arg_utils.py [hunk 2]
```diff
@@ -39,7 +39,7 @@ class EngineArgs:
     quantization: Optional[str] = None
     enforce_eager: bool = False
     max_context_len_to_capture: int = 8192
-    disable_custom_all_reduce: bool = False
+    disable_custom_all_reduce: bool = True
```
Comment (WoosukKwon) [inline vllm/engine/arg_utils.py:42]: Can we keep this `False` by default? We'd like to minimize the number of knobs that users have to tune. Also, this fix does not allow our CLI users (e.g., for OpenAI server) to enable the custom all reduce.

File: vllm/model_executor/parallel_utils/custom_all_reduce.py [hunk 3]
```diff
@@ -134,18 +141,48 @@ def _is_full_nvlink(rank, world_size):
 
 
 def _can_p2p(rank: int, world_size: int) -> bool:
+    num_dev = torch.cuda.device_count()
+    # note: num dev can be larger than world_size if we're only using
+    # first few GPUs
```
Comment (WoosukKwon) [inline vllm/model_executor/parallel_utils/custom_all_reduce.py:146]: This is the case when more than one nodes (hosts) are used for TP, right?

File: csrc/custom_all_reduce_test.cu
Comment (WoosukKwon) [inline csrc/custom_all_reduce_test.cu:1]: Can we move this test to Python so that we can easily run it?

**Review by WoosukKwon — APPROVED** [review]
@hanzhi713 Apologies for the very late review. I had no bandwidth recently.

The PR looks good to me, while I still don't understand the cause of the bug. I let some minor comments. Please take a look at them.

File: vllm/model_executor/parallel_utils/custom_all_reduce.py (hunk 3 above)
Comment (hanzhi713) [reply vllm/model_executor/parallel_utils/custom_all_reduce.py:146]: No. This check is for the case that the users have N GPUs, but are only using the first M GPUs where M < N.

File: vllm/engine/arg_utils.py (hunk 2 above)
Comment (hanzhi713) [reply vllm/engine/arg_utils.py:42]: Sure

**Comment by WoosukKwon** [comment]
@hanzhi713 Thanks! Let me test the PR and merge.

**Comment by WoosukKwon** [comment]
@hanzhi713 I will merge this as I found this worked well on my 4 A100-80GB machine. Thanks for the fix!

**Comment by WoosukKwon** [comment]
@hanzhi713 Actually, we recently found that TRT-LLM's custom all reduce kernel is extremely simple. Do you have an idea why it can be much simpler than this implementation? What do you think about using TRT-LLM's kernel?

**Comment by hanzhi713** [comment]
> @hanzhi713 Actually, we recently found that TRT-LLM's custom all reduce kernel is extremely simple. Do you have an idea why it can be much simpler than this implementation? What do you think about using TRT-LLM's kernel?

It looks to me the implementation has about the same complexity as mine. What makes you think it looks simpler?

**Comment by garycaokai** [comment]
in my test, 4 non NVLink-capable GPUs situation. custom allreduce have 20% performance improvement when batch 1.
  // for 4 or more non NVLink-capable GPUs, custom allreduce provides little
  // performance improvement over NCCL.

**Comment by garycaokai** [comment]
> in my test, 4 non NVLink-capable GPUs situation. custom allreduce have 20% performance improvement when batch 1. // for 4 or more non NVLink-capable GPUs, custom allreduce provides little // performance improvement over NCCL.

@hanzhi713 can you open 4 non NVLink-capable GPUs situation as an option?

**Comment by hanzhi713** [comment]
@garycaokai Curious about your setup. What GPUs are you using? Are they all connected to a PCIe switch or are they connected to CPU directly?

Yes I can provide that as an option if I find some time to work on this.

**Comment by hanzhi713** [comment]
@garycaokai Also, how did you measure the performance improvement? Is it a latency benchmark? And what is the configuration for the benchmark?

**Comment by garycaokai** [comment]
> @garycaokai Curious about your setup. What GPUs are you using? Are they all connected to a PCIe switch or are they connected to CPU directly?
> 
> Yes I can provide that as an option if I find some time to work on this.
72b int4 tp=4, 1 batch,  decode speed from 20 token/s to 26 token/s
A30 * 8 gpus .
#nvidia-smi topo -m
	GPU0	GPU1	GPU2	GPU3	GPU4	GPU5	GPU6	GPU7	NIC0	NIC1	CPU Affinity	NUMA Affinity	GPU NUMA ID
GPU0	 X 	PIX	PIX	PIX	SYS	SYS	SYS	SYS	NODE	NODE	0-11,48-59	0		N/A
GPU1	PIX	 X 	PIX	PIX	SYS	SYS	SYS	SYS	NODE	NODE	0-11,48-59	0		N/A
GPU2	PIX	PIX	 X 	PIX	SYS	SYS	SYS	SYS	NODE	NODE	0-11,48-59	0		N/A
GPU3	PIX	PIX	PIX	 X 	SYS	SYS	SYS	SYS	NODE	NODE	0-11,48-59	0		N/A
GPU4	SYS	SYS	SYS	SYS	 X 	PIX	PIX	PIX	SYS	SYS	24-35,72-83	2		N/A
GPU5	SYS	SYS	SYS	SYS	PIX	 X 	PIX	PIX	SYS	SYS	24-35,72-83	2		N/A
GPU6	SYS	SYS	SYS	SYS	PIX	PIX	 X 	PIX	SYS	SYS	24-35,72-83	2		N/A
GPU7	SYS	SYS	SYS	SYS	PIX	PIX	PIX	 X 	SYS	SYS	24-35,72-83	2		N/A
NIC0	NODE	NODE	NODE	NODE	SYS	SYS	SYS	SYS	 X 	PIX				
NIC1	NODE	NODE	NODE	NODE	SYS	SYS	SYS	SYS	PIX	 X 				

Legend:

  X    = Self
  SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)
  NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
  PXB  = Connection traversing multiple PCIe bridges (without traversing the PCIe Host Bridge)
  PIX  = Connection traversing at most a single PCIe bridge
  NV#  = Connection traversing a bonded set of # NVLinks

NIC Legend:

  NIC0: mlx5_0
  NIC1: mlx5_1

**Comment by hanzhi713** [comment]
I see this makes a lot of sense. Usually A30 machines have GPUs connected to CPU directly, and CPUs are often terrible PCIe switches. My implementation relying PCIe P2P. However, in your case you have a PCIe switch connecting to each group of 4 GPUs. Given the much better switching performance, my implementation may work. 

Let me see if I can find machines of this topology

**Pull request merged**